src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderStatic.cs

```diff
@@ -0,0 +1,176 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Core.Data;
+using Microsoft.ML.Data.StaticPipe.Runtime;
+using Microsoft.ML.Data.StaticPipe;
+using Microsoft.ML.Runtime.Internal.Utilities;
+
+namespace Microsoft.ML.Runtime.Data
+{
+    public sealed partial class TextLoader
+    {
+        public static DataReader<IMultiStreamSource, TTupleShape> CreateReader<TTupleShape>(
```

Comment:
From an API perspective, it seems surprising to me that `TextLoader.CreateReader` returns a `DataReader`.  I guess I would have expected it to return a `TextLoader`.  Is this a pattern to be used by all "loaders"/"readers".

Which kind of asks another question: What's the difference between a "loader" and a "reader"?

Reply:
A loader is an old concept which will almost certainly be deprecated. A "reader" (name is not finalized) is a new concept. In the old world where this was a tool and not an API, the idea of the head of a pipeline being *assumed* to come from file(s) (as represented by `IMultiStreamSource`) was a priviledged and central concept -- and this was called a loader. Now @Zruty0 has generalized this notion and the head of a pipeline is something called a reader, with a generic input parameter which may or may not be an `IMultiStreamSource`, so he had to come up with a different name, which is called "reader."

The following will happen. We will replace `IDataLoader` altogether with some sort of concept that takes a stream source -- that successor concept is, for the moment, termed `IDataReader<IMultiStreamSource>`. (We may also choose to rename what we are temporarily calling a "reader," as well. Who knows. Maybe we'll name it Fred.) So when you see loader you are seeing something whose name will I think be changed. (At least, I would not welcome these two concepts being present in the code in the same time.)

At any rate I view this as merely a renaming question. What is perhaps a more urgent question is whether we got the *shape* of the API right, rather than its names. In particular, in this case, there were *three* choices we had when we created this object, that it would be a data reader estimator, or a data reader, or just the data. We chose data reader for various reasons. Is that itself confusing?

Reply:
Maybe it's sort of like this. Imagine hypothetically there was some early version of .NET that had this idea of an `Iterable` as a central concept that involved iteration over integers specifically. But then some bright developer had the idea, "heck, what if I want to iterate over something else, like floats, or even some sort of reference type." So then they replace this old concept of `IIterable` with `IEnumerable<int>`. But, realistically, the pace of development being what it is *possibly* that cannot reasonably happen all at once. That is sort of how I view `IDataLoader` and `IDataReader<>`.

We're moving towards the generalization, but it's not something that will happen all at once. (And certainly not in this PR -- I am introducing here PiGSTy, and PiGSTy is deliberately engineered in such a way as to not change the paradigms of the underlying dynamic system, it merely makes them less hazardous to work with while cleaving to their idioms as closely as possible -- I don't even want to diverge as far as using a different name for an analogous method if I can possibly help it.)